engineering/docs/go_conventions.md

```diff
@@ -0,0 +1,139 @@
+
+## General
+
+### Logging
+
+Use `github.com/solo-io/go-utils/contextutils` for logging. 
+Initialize a context with useful information, including a version field.
+
+```go
+func getInitialContext() context.Context {
+	loggingContext := []interface{}{"version", version.Version}
+	ctx := contextutils.WithLogger(context.Background(), BotName)
+	ctx = contextutils.WithLoggerValues(ctx, loggingContext...)
+	return ctx
+}
+```
+
+Log using zap fields where possible. If a number of fields are added, break the line into a single
+ field per line: For instance:
+
+```go
+contextutils.LoggerFrom(ctx).Infow("Downloading repo archive",
+			zap.String("owner", r.owner),
+			zap.String("repo", r.repo),
+			zap.String("sha", r.sha))
+```
+
+### Interfaces
+
+Define interfaces when writing new utilities/modules that allow for mocking out dependencies and unit testing. 
+The default struct implementation for an interface should be have the same name, with a lowercase first letter. 
+A constructor should be used for instantiation. The interface, struct, and constructor should be near the top of a file. 
+
+```go
+type ChangelogReader interface {
+	GetChangelogForTag(ctx context.Context, tag string) (*Changelog, error)
+	ReadChangelogFile(ctx context.Context, path string) (*ChangelogFile, error)
+}
+
+type changelogReader struct {
+	code vfsutils.MountedRepo
+}
+
+func NewChangelogReader(code vfsutils.MountedRepo) ChangelogReader {
```

Comment:
I was recently made aware by @yuval-k of the "_accept interfaces, return structs_" best practice. There are different blogs which describe the advantages, but the basic idea is that it's the caller who should be able to define the interface it needs. I think it is a good rule of thumb and we should consider adopting it.

Reply:
I was under the (mistaken) impression that we couldn't return lower-cased structs and have them be accessible to other packages. I'm still not sure there's much of a difference to returning structs vs. interfaces in this context, but I'm happy to switch to that.